### PR TITLE
Refactoring of API Responses

### PR DESCRIPTION
**Pull request by alexbowers**
All controllers can now just `return` and it will be given the correct structure, and a `200` response code.

Any different responses can be achieved by throwing the relevant exception `ApiException`.

Uses of throwing that exception can look like:

```php
$v = Validator::make($this->data, $this->rules);

if (!$v->passes()) {
    throw (new ApiException("Failed to pass validation", 400))->withErrors($v);
}
```

```php
throw (new ApiException("Failed to pass validation", 400))->withErrors($v);
```

Will return the errors from the validation for use in the front-end.

```php
throw new ApiException("Failed to pass validatiion", 400);
```

Will throw the error exception back to the frontend, with an empty error set, but with the Validation Message.

Still heavily work in progress, this is to show how the PHP side of code will look / work.

JS needs to be fixed to match new expectation (data isn't immediately available anymore, you have to go into `response.data.message` now.

### CODE REVIEW

**Comment by syropian** [comment]
Some minor cs issues it looks like, but otherwise 💯

File: app/Http/Middleware/Response.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+     * @param  \Illuminate\Http\Request $request
+     * @param  \Closure $next
+     *
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        $response = $next($request);
+
+        $status_code = $response->getStatusCode();
```
Comment (joshhornby) [inline app/Http/Middleware/Response.php]: Can we use camelCase for variables?

File: app/Http/Middleware/Response.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        $content = $response->getContent();
+
+        if (empty( $content )) {
+            $content = "Success, but no response received.";
+        }
+
+        return is_json($content) ? json_decode($content) : $content;
+    }
+
+    private function code_405(BaseResponse $response) : string
```
Comment (joshhornby) [inline app/Http/Middleware/Response.php]: same again, can this be camelCase?

File: bootstrap/app.php [hunk 3]
```diff
@@ -41,6 +41,11 @@
     Astral\Exceptions\Handler::class
 );
 
+/**
+ * Some helper functions.
+ */
+require "helpers.php";
```
Comment (joshhornby) [inline bootstrap/app.php]: I really don't like use global functions like this, much rather have a class or a trait.

In this case we only use the code once so why not a `protected` function?

File: app/Exceptions/ApiException.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+class ApiException extends \Exception
+{
+    protected $message;
+    protected $code;
+    protected $previous;
+    /**
+     * @type \Illuminate\Support\MessageBag
+     */
+    protected $errors;
```
Comment (joshhornby) [inline app/Exceptions/ApiException.php]: Can we use this format instead

```
/** @var \Illuminate\Support\MessageBag */
protected $errors;
```

File: app/Http/Controllers/TagController.php [hunk 5]
```diff
@@ -52,8 +50,7 @@ public function reorder(Request $request)
             $tag->sort_order = $row['sort_order'];
             $tag->save();
         }
-        $tags = Tag::with('stars.tags')->where('user_id', Auth::id())->orderBy('sort_order', 'asc')->get();
-        return response()->json(compact('tags'), 200);
+        return Tag::with('stars.tags')->where('user_id', Auth::id())->orderBy('sort_order', 'asc')->get();
```
Comment (joshhornby) [inline app/Http/Controllers/TagController.php]: Every return should have a blank line before.

File: app/Http/Middleware/Response.php (hunk 1 above)
Comment (syropian) [inline app/Http/Middleware/Response.php]: Agreed, I tend to prefer camelCase as well.

Comment (alexbowers) [inline app/Http/Middleware/Response.php]: Okay, i'll refactor to be camel cased. This is gonna be a weird habit for me to get used to :P

File: bootstrap/app.php (hunk 3 above)
Comment (alexbowers) [inline bootstrap/app.php]: This is generally a pretty useful function to have.

I agree that having global functions is messy, but this is following on from Laravel\Support\helpers.php own structure, where they provide some globally helpful functions.

Having this inside of a class seems pointless, but could be done if necessary.

File: app/Exceptions/ApiException.php (hunk 4 above)
Comment (alexbowers) [inline app/Exceptions/ApiException.php]: I much prefer docblocks to always be over a minimum of three lines, makes it more readable IMO. @syropian input?

File: app/Http/Controllers/TagController.php (hunk 5 above)
Comment (alexbowers) [inline app/Http/Controllers/TagController.php]: Okay. can the phpcs fixer define this too to help catch issues?

**Comment by alexbowers** [comment]
Tasks to do:

- [ ] Space before all return's
- [ ] Come to agreement on property docblock 1 or 3 lines.
- [ ] Come to agreement on whether global helper is_json is better placed in a class / trait.
- [ ] Camelcase the variables everywhere

File: bootstrap/app.php (hunk 3 above)
Comment (joshhornby) [inline bootstrap/app.php]: Global function IMO are the worst thing about Laravel. They encourage bad design and lazy programming.

For now I'd add this as a simple function where we need it, if we need to use it later on we could always extend the `Str` class as I believe that would be a logical place for it to live.

File: app/Exceptions/ApiException.php (hunk 4 above)
Comment (joshhornby) [inline app/Exceptions/ApiException.php]: I disagree, take the two examples:

```
    /**
     * @var string
     */
    protected $message;

    /**
     * @var int
     */
    protected $code;

    /**
     * @var string
     */
    protected $previous;

    /**
     * @var \Illuminate\Support\MessageBag
     */
    protected $errors;
```
vs
```
    /** @var string */
    protected $message;

    /** @var int */
    protected $code;

    /** @var string */
    protected $previous;

    /** @var \Illuminate\Support\MessageBag */
    protected $errors;
```

We have just added 8 extra lines to the file for no reason.

File: app/Http/Controllers/TagController.php (hunk 5 above)
Comment (joshhornby) [inline app/Http/Controllers/TagController.php]: Yes, my new PR catches this.

**Comment by syropian** [comment]
@alexbowers Any idea when you can finish this up? No huge rush, just trying to plan my week 😅

**Comment by alexbowers** [comment]
I'll have an opportunity to do some more work on it this evening, I'll need
to do a rebase then I should be able to either complete or get it close
tonight

File: app/Http/Middleware/Response.php (hunk 2 above)
Comment (alexbowers) [inline app/Http/Middleware/Response.php]: What do you expect to be camel cased in here?

File: bootstrap/app.php (hunk 3 above)
Comment (alexbowers) [inline bootstrap/app.php]: I'll keep this as is for now. We can change and refactor it at a later date away from global functions.

File: app/Http/Middleware/Response.php (hunk 2 above)
Comment (joshhornby) [inline app/Http/Middleware/Response.php]: function names `code200` etc

File: bootstrap/app.php (hunk 3 above)
Comment (joshhornby) [inline bootstrap/app.php]: Can we just pull it into a `private` function? No point starting to add global functions just to take them away again in the not so distant future.

**Comment by joshhornby** [comment]
Once this PR lands I am going to take a look at caching.

**Comment by alexbowers** [comment]
Are you able to hop onto gitter? Few Q's for you.

**Comment by joshhornby** [comment]
Guess this is more a question for @syropian but why do we include the `public/js/app.bundle.js` file? This PR now shows 66,000 lines changed!!

Not sure what the correct way is, but including this on every JS update seems wrong.

**Comment by joshhornby** [comment]
PHP looks good to go though :+1:

**Comment by syropian** [comment]
@joshhornby the bundle currently gets checked in because I use Git for deployments via Laravel Forge. I could potentially remove it from version control and try and get the server to generate the bundle on deploy. Let me know your thoughts.

**Comment by alexbowers** [comment]
I think bundling it is fine, But it needs to be consistent, I dont know what I was doing wrong where i was not getting consistent bundles, with version constants changing etc.

**Comment by syropian** [comment]
@alexbowers I just realized that we're checking in `public/js/app.bundle.js` instead of the files in the `public/build` folder (which are getting gitignore'd). We should probably fix that  😅

**Comment by syropian** [comment]
@alexbowers I've checked `public/build` into Git now, so versioned assets are under version control now.

**Comment by alexbowers** [comment]
Nice. I still dont trust the building, for me things like version constants were changing, which in my mind shouldn't. Are you sure the build works properly, and that the command used to build is just `gulp --production`

**Comment by syropian** [comment]
Ah I haven't been using the production flag before pushing to Git. Makes sense to do that though. I'll do a production build and push shortly. That way the difference shouldn't be as big.

**Comment by joshhornby** [comment]
I always run `gulp --production`, guess thats all we need to do on the PR and then it can be merged.

**Comment by syropian** [comment]
Cool, yeah I mean I run `gulp watch` when I'm deving, but when you're ready to file a PR, `gulp --production` should get run. 👍

**Comment by alexbowers** [comment]
@syropian This has now been 'gulped'

**Comment by syropian** [comment]
Great! Is it ready to go?

**Comment by alexbowers** [comment]
I think so, but a check by someone else wouldn't go amiss

**Comment by syropian** [comment]
@joshhornby If you can give this a last once-over, that would be much appreciated 🙏 

@alexbowers Btw, I sent you an invite to the Astral organization, which gives you write permissions on the base repo, so you don't need to run off a forked copy.

**Comment by joshhornby** [comment]
@syropian +1 from me

**Comment by alexbowers** [comment]
Merged.

**Comment by syropian** [comment]
🎉

**Pull request merged**